crates/libs/core/src/sync/cancel.rs

```diff
@@ -0,0 +1,729 @@
+#[implement(IFabricAsyncOperationContext)]
+pub struct BridgeContext3<T>
+where
+    T: 'static,
+{
+    content: Cell<Option<T>>,
+    is_completed: Cell<bool>,
+    is_completed_synchronously: bool,
+    callback: IFabricAsyncOperationCallback,
+    token: CancellationToken,
+}
+
+impl<T> BridgeContext3<T>
+where
+    T: Send,
+{
+    fn new(callback: IFabricAsyncOperationCallback, token: CancellationToken) -> Self {
+        Self {
+            content: Cell::new(None),
+            is_completed: Cell::new(false),
+            is_completed_synchronously: false,
+            callback,
+            token,
+        }
+    }
+
+    /// Creates the context from callback, and returns a cancellation token that
+    /// can be used in rust code, and the cancellation token is hooked into self,
+    /// where Cancel() api cancels the operation.
+    pub fn make(callback: Option<&IFabricAsyncOperationCallback>) -> (Self, CancellationToken) {
+        let token = CancellationToken::new();
+        let ctx = Self::new(callback.unwrap().clone(), token.clone());
+        (ctx, token)
+    }
+
+    /// Spawns the future on rt.
+    /// Returns a context that can be returned to SF runtime.
+    /// This is intended to be used in SF Begin COM api, where
+    /// rust code is spawned in background and the context is returned
+    /// to caller.
+    /// This api is in some sense unsafe, because the developer needs to ensure
+    /// the following:
+    /// * return type of the future needs to match SF COM api end return type.
+    pub fn spawn<F>(
+        self,
+        rt: &impl Executor,
+        future: F,
+    ) -> crate::Result<IFabricAsyncOperationContext>
+    where
+        F: Future<Output = T> + Send + 'static,
+    {
+        let self_cp: IFabricAsyncOperationContext = self.into();
+        let self_cp2 = self_cp.clone();
+        rt.spawn(async move {
+            let ok = future.await;
+            let self_impl: &BridgeContext3<T> = unsafe { self_cp.as_impl() };
+            self_impl.set_content(ok);
+            self_impl.set_complete();
+            let cb = self_impl.Callback().unwrap();
+            unsafe { cb.Invoke(&self_cp) };
+        });
+        Ok(self_cp2)
+    }
+
+    /// Get the result from the context from the SF End COM api.
+    /// This api is in some sense unsafe, because the developer needs to ensure
+    /// the following:
+    /// * context impl type is `BridgeContext3`, and the T matches the SF end api
+    /// return type.
+    /// Note that if T is of Result<ICOM> type, the current function return type is
+    /// Result<Result<ICOM>>, so unwrap is needed.
+    pub fn result(context: Option<&IFabricAsyncOperationContext>) -> crate::Result<T> {
+        let self_impl: &BridgeContext3<T> = unsafe { context.unwrap().as_impl() };
+        self_impl.consume_content()
+    }
+
+    // TODO: send and comsume is expected to happend accross threads.
+    // Even though we use a oneshot channel to send the signal,
```

Comment:
Oneshot channel should already have a memory barrier internally (you can confirm if you'd like), so this is ok as is

Reply:
This might be independent of oneshot. We set the result in the ctx in one thread, and then trigger the SF callback to notify the result is ready, and then the SF might use another thread to retrieve the result from ctx. 

The comment might have been copied from earlier BridgeContext implementation, which might have relied on oneshot.
The original comment might also indicate that if the Bridge is used in the Proxy layer in rust, then the Proxy layer has an oneshot channel  (which is in the SF callback), then we are covered. The only missing case are the SF callbacks which do not have any memory barriers inside.

I will keep the comment as is for now.